runtime/integration-tests/src/generic/envs/runtime_env.rs

```diff
@@ -0,0 +1,200 @@
+			let signature =
+				MultiSignature::Sr25519(raw_payload.using_encoded(|payload| who.sign(payload)));
+
+			let multi_address = (Address::Id(who.to_account_id()), signature, signed_extra);
+
+			let extrinsic =
+				<T::Block as Block>::Extrinsic::new(runtime_call, Some(multi_address)).unwrap();
+
+			self.nonce += 1;
+
+			T::apply_extrinsic(extrinsic).unwrap()
+		})
+	}
+
+	fn pass(&mut self, blocks: Blocks<T>) {
+		self.ext.borrow_mut().execute_with(|| {
+			let next = frame_system::Pallet::<T>::block_number() + 1;
+
+			let end_block = match blocks {
+				Blocks::ByNumber(n) => next + n,
+				Blocks::BySeconds(secs) => {
+					let blocks = secs / pallet_aura::Pallet::<T>::slot_duration();
+					if blocks % pallet_aura::Pallet::<T>::slot_duration() != 0 {
+						blocks as BlockNumber + 1
+					} else {
+						blocks as BlockNumber
+					}
+				}
+				Blocks::UntilEvent { limit, .. } => limit,
+			};
+
+			for i in next..end_block {
+				T::finalize_block();
+				Self::prepare_block(i);
+
+				if let Blocks::UntilEvent { event, .. } = blocks.clone() {
+					let event: T::RuntimeEventExt = event.into();
+					if frame_system::Pallet::<T>::events()
+						.into_iter()
+						.find(|record| record.event == event)
+						.is_some()
+					{
+						break;
+					}
+				}
+			}
+		})
+	}
+
+	fn state_mut<R>(&mut self, f: impl FnOnce() -> R) -> R {
+		self.ext.borrow_mut().execute_with(f)
+	}
+
+	fn state<R>(&self, f: impl FnOnce() -> R) -> R {
+		self.ext.borrow_mut().execute_with(|| {
+			transactional::with_transaction(|| {
+				// We revert all changes done by the closure to offer an inmutable state method
+				TransactionOutcome::Rollback::<Result<R, DispatchError>>(Ok(f()))
+			})
+			.unwrap()
+		})
+	}
+}
+
+impl<T: Runtime> RuntimeEnv<T> {
+	fn prepare_block(i: BlockNumber) {
+		let slot = Slot::from(i as u64);
+		let digest = Digest {
+			logs: vec![DigestItem::PreRuntime(AURA_ENGINE_ID, slot.encode())],
+		};
+
+		let header = Header {
+			number: i,
+			digest,
+			state_root: H256::default(),
+			extrinsics_root: H256::default(),
+			parent_hash: H256::default(),
+		};
+
+		T::initialize_block(&header);
+
+		let timestamp = i as u64 * pallet_aura::Pallet::<T>::slot_duration();
+		let inherent_extrinsics = vec![
+			Extrinsic::new(Self::cumulus_inherent(i), None).unwrap(),
+			Extrinsic::new(Self::timestamp_inherent(timestamp), None).unwrap(),
+		];
+
+		for extrinsic in inherent_extrinsics {
+			T::apply_extrinsic(extrinsic).unwrap().unwrap();
+		}
+	}
+
+	fn cumulus_inherent(i: BlockNumber) -> T::RuntimeCall {
+		let mut inherent_data = InherentData::default();
+
+		let sproof_builder = RelayStateSproofBuilder::default();
+		let (relay_parent_storage_root, relay_chain_state) =
+			sproof_builder.into_state_root_and_proof();
+
+		let cumulus_inherent = ParachainInherentData {
+			validation_data: PersistedValidationData {
+				parent_head: vec![].into(),
+				relay_parent_number: i,
```

Comment:
Nit: I think this might become problemativ once we rely on the relay chain block number in some code locations. But not the case right now. But we will probably switch to the relay-chain as a our clock sooner than later.

Reply:
>  once we rely on the relay chain block number in some code locations.

Makes sense, taken into account!